### PR TITLE
#171380745 Signup email verification

### PR DESCRIPTION
**Pull request by ricahcyuzuzo**
## What does this PR do?
Send Verification email upon a successful SignUp submission
## Description of Task to be Completed?
- Sending Verification email via running signup endpoint
## How should this Manually tested?
- First, clone this repository by `https://github.com/Stackup-Rwanda/stackup2-barefoot-backend.git`
- fetch this branch
- In the file .env file include the following Variabes: 
     - `SENDGRID_API_KEY=your_api_key`
     - `APP_URL=the link of the host you are using` _eg: http://localhost:3000_ 
     - `SENT_EMAIL=your_email`. This is the email you will use to send email
- Go to your terminal from the local git repo directory and hit `npm run dev`.
- run the following endpoint with postman.
- `/api/auth/signup`.
#### This is the body

> {
> "firstName": "John",
> "lastName": "Doe",
> "email": "john@doe.com",
> "password": "your_password",
> "gender": "Male",
> "address": "kigali"
> }
After Signing up, go to your email and click on a button to comfirm your your account 
## Any Background Context you want to provide?
- You should create the account on [www.sendgrid.com](https://sendgrid.com/) to get a key
## What are relevant Pivotal Tracker stories?
- [#171380745](https://www.pivotaltracker.com/story/show/171380745)
## Screenshots (If appropriate)
The screenshot of the email
Screenshot from 2020-03-11 20-10-03

### CODE REVIEW

**Comment by the22mastermind** [comment]
@ricahcyuzuzo You did a good job but I would like to suggest that don't make a route for sending email. Instead you could rename that emailVerification.service.js to sendEmail.service.js and inside you create a class, call it EmailSender {} and inside that class you put your sendMail function and rename it sendSignUpVerificationLink().
In signUp function after generating the token, you could then call your function to send the email, this way every successful signup will have an email sent to the user with that verification link. And also the EmailSender class could be used to send other types of emails.
So to test your PR, one would just test the signup.
Lastly, make sure you resolve the conflicts.

**Review by descholar-ceo — COMMENTED** [review]
1. Could you add more details in *How should this Manually tested?*

**Review by since9629 — CHANGES_REQUESTED** [review]
@ricahcyuzuzo could you fix CI issues

**Comment by ricahcyuzuzo** [comment]
> @ricahcyuzuzo could you fix CI issues

@karamuka It is solved!

**Comment by ricahcyuzuzo** [comment]
> 1. Could you add more details in _How should this Manually tested?_

@descholar-ceo Done!!

**Review by since9629 — APPROVED** [review]

**Review by Ugizwenayo-Divine — APPROVED** [review]

File: .env.example [hunk 1]
```diff
@@ -26,3 +26,8 @@ PORT=
 SECRET=
 JWT_KEY=YOUR_SECRET_KEY_GOES_HERE
 SIGN_UP_TOKEN_EXPIRES_IN=VALUE #Eg 1h, 24h, 30d
+
+#Sending email
+MY_DEV_KEY=YOUR_SENDGRID_KEY_FOR_DEVELOPMENT
+MY_TESTING_KEY=YOUR_SENDGRID_KEY_FOR_TESTING
+MY_EMAIL=barefootnomad2@gmail.com
```
Comment (descholar-ceo) [inline .env.example]: could you solve this newline issue?

File: package.json [hunk 2]
```diff
@@ -8,7 +8,7 @@
     "dev": "nodemon --exec babel-node ./src/index.js",
     "migrations": "npx sequelize-cli db:migrate",
     "undo": "npx sequelize-cli db:migrate:undo",
-    "test": "npm run migrations && nyc mocha --require @babel/register ./tests/*.js  --timeout 25000 --exit && npm run undo",
+    "test": "npm run migrations && nyc mocha --require @babel/register ./tests/*.js  --timeout 25000 --exit",
```
Comment (descholar-ceo) [inline package.json]: why did you change this script?

**Review by descholar-ceo — COMMENTED** [review]
Left you comments in codes

File: package.json (hunk 2 above)
Comment (ricahcyuzuzo) [reply package.json]: Oh It's my bad!

**Review by descholar-ceo — APPROVED** [review]

File: src/services/sendEmail.service.js [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+      product: {
+        name: 'Barefoot Nomad',
+        link: '#'
+      }
+    });
+
+    const generateEmail = async () => ({
+      body: {
+        name: 'Successful signup verification email',
+        intro: 'Congractulations, you have sucessfully created the account!',
```
Comment (descholar-ceo) [inline src/services/sendEmail.service.js]: I think you wanted to write `Congratulation`

File: src/controllers/authentication.controller.js [hunk 4]
```diff
@@ -32,6 +34,8 @@ export default class AuthenticationController {
       const saveUser = await handleSignUp(userData);
       const savedUserObject = _.omit(saveUser, 'password');
       const token = await generateToken(savedUserObject);
+      console.log(req.body.email);
+      await sendEmail.sendSignUpVerificationLink(req.body.email, `${process.env.APP_URL}`);
       return successResponse(res, statusCodes.created, customMessages.userSignupSuccess, token);
```
Comment (descholar-ceo) [inline src/controllers/authentication.controller.js:57]: As you send the email after successful registration, don't you think that you should edit the message returned on this signup API? Because after a user has successfully created the account see on the screenshot the mockup of the page where he/she will be redirected 
Screenshot from 2020-03-11 14-30-32
that is the page where a user will be redirected after successful registration, but as I see you are still sending the token key and the message on the API 
Screenshot from 2020-03-11 14-32-17
that is what I got after trying to signup I suggest you change this signup message a user receives after successful signup. It should match what is on the mockup.

**Review by descholar-ceo — CHANGES_REQUESTED** [review]
Please consider my comments in your codes and do some work on them

File: src/services/sendEmail.service.js (hunk 3 above)
Comment (ricahcyuzuzo) [reply src/services/sendEmail.service.js]: @muchristian You got it wrong. that is the template that will be sent to the Email

**Review by Muhire-Josue — APPROVED** [review]
LGTM

**Review by descholar-ceo — APPROVED** [review]
